[Blackjack - Step1] Submit Deniz' mission

deniz-oezdemir opened the pull request:
### PR Ready for Review

Hi there,

I hope you are doing well!

This PR is ready for review. I'd appreciate your thoughts on a few questions that came up while working on it:

1. **Naming inside `Players` class**
   Currently, we refer to the list of players as `players`. Would something else be a clearer for this collection?

2. **Number of decks**
   Should we simulate multiple decks, like in a real casino? Right now, it's a single deck. Using more might better reflect how Blackjack is usually played.
   Anyways, like in most casinos we have capped the maximum amount of players at 7 so we will not run out of cards with one deck.

3. **Controller state**
   We keep some private variables in the `Controller` to track state. Is this an acceptable tradeoff, or should we use a more declarative approaches and pass variables between functions instead?

4. **`toString()` in `ResultsEvaluator` data classes**
   We’ve overloaded `toString()` to make it easier to print results. Should we instead move formatting into a separate function inside `OutputView`?

Additionally, if you spot anything that could be written in a more Kotlin-idiomatic way, I’d really value that feedback.

Thanks in advance for your time. All tips and suggestions are much appreciated.

Best,
Deniz

petrabenczedh committed:
docs(readme): initial features

petrabenczedh committed:
init(blackjack): folder structure

petrabenczedh committed:
feat(blackjack): create Participant interface and Player class, test them

petrabenczedh committed:
feat(Controller): read player names and init players

petrabenczedh committed:
feat(OutputView): print player names

petrabenczedh committed:
feat(Card): create card class with rank and suit

petrabenczedh committed:
feat(Suit, Rank): create enums for card rank and suit

petrabenczedh committed:
feat(CardDeck): create CardDeck and test it

petrabenczedh committed:
feat(CardDeck, Participant): add draw card and hit functions

petrabenczedh committed:
feat(Dealer): new class

petrabenczedh committed:
feat(Controller): print handcards of players

petrabenczedh committed:
style(): format files with autoformatting

petrabenczedh committed:
feat(Controller): deal cards for players

petrabenczedh committed:
feat(HandCards, Player): calculate total sum and show if player is busted

petrabenczedh committed:
docs(README): update features

petrabenczedh committed:
test(Dealer): test drawing logic for dealer

petrabenczedh committed:
test(Dealer): test drawing logic for dealer

petrabenczedh committed:
feat(OutputView): display cards and totals summary

petrabenczedh committed:
refactor(HandCards): add type to properties

petrabenczedh committed:
docs(README)

petrabenczedh committed:
feat(Participant): add isBusted function

petrabenczedh committed:
feat(Players): add size property

petrabenczedh committed:
feat(ResultEvaluator): add class to calc results

petrabenczedh committed:
feat(OutputView): print results

petrabenczedh committed:
feat(ResultEvaluator): calculate results for players

petrabenczedh committed:
style(): minor changes

petrabenczedh committed:
style(): minor changes

petrabenczedh committed:
refactor(ResultEvaluator, HandCards): instead of properties use functions, operator overload in HandCards

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
update(gitignore)

petrabenczedh committed:
update(gitignore)

petrabenczedh committed:
docs(README): questions

petrabenczedh committed:
refactor(Participant): use abstract class instead of interface for Participant

petrabenczedh committed:
refactor(ErrorMessage, PromptMessage): separate messages from core logic

petrabenczedh committed:
test(ResultEvaluator)

petrabenczedh committed:
style(OutputView)

petrabenczedh committed:
style(InputView)

petrabenczedh committed:
fix(OutputView): add new line before printing totals

petrabenczedh committed:
style(README):

deniz-oezdemir committed:
feat(gitignore): update gitignore

deniz-oezdemir committed:
Merge remote-tracking branch 'upstream/step1' into step1

deniz-oezdemir committed:
style(ResultEvaluator): remove commented pseudocode

pkeugine commented on README.md:
```diff
@@ -1 +1,93 @@
-# kotlin-blackjack
\ No newline at end of file
+# kotlin-blackjack
+Implementation of the Blackjack card game in Kotlin.
```
Nice work on organizing README and stacking commits 👍 

It'll be much better if we could add titles to some commits that don't have it,
and name them so that it reflects what that commits contain.
For example, there are two commits named `test(Dealer): test drawing logic for dealer`, but only one of them contains the actual test.

<img width="282" height="50" alt="Screenshot 2025-07-11 at 11 48 58 AM" src="https://github.com/user-attachments/assets/e9689ddc-2774-49f2-8b96-6a2052fee6f5" />

deniz-oezdemir replied:
Yes, one of those was by mistake. Good catch! :-)

pkeugine commented on src/test/kotlin/study/DSLTest.kt:
```diff
@@ -0,0 +1,34 @@
+ * }
+ */
+
+class DSLTest {
```
This test file is not allowing me to run all the tests at once.
I think you've tried to remove this file in`.gitignore`, but it slipped into this PR.
Please remove this file. :)

deniz-oezdemir replied:
Good catch, I have deleted the study test file and the respective include in the `.gitignore`.

pkeugine commented on src/main/kotlin/blackjack/model/Dealer.kt:
```diff
@@ -0,0 +1,15 @@
+) : Participant() {
+
+    fun shouldDraw(): Boolean {
+        return handCards.total <= 16
```
I think we can give a name to the number 16, and treat it as a magic number.
Same goes to number 21, since it's a special number in blackjack.
How does this sound?

deniz-oezdemir replied:
Thanks for the suggestion!
I’ve extracted the magic numbers into a single `Rules` singleton:

pkeugine commented on src/main/kotlin/blackjack/model/Players.kt:
```diff
@@ -0,0 +1,18 @@
+package blackjack.model
+
+class Players(val players: List<Player>) {
```
**Q. Currently, we refer to the list of players as players. Would something else be a clearer for this collection?**

What were the other options that you and your pair thought of?
I'd like to know that first. :)

deniz-oezdemir replied:
We thought of two options:
(1) Rename the class to "Table"
(2) Rename the attribute to "seats" or "spots"

pkeugine replied:
Well, unless there's a special reason behind it (let's say for example, the company or the game system specifically uses that term), I think `Players` is the most intuitive option. :)

deniz-oezdemir replied:
Cool, then I leave it as is for now. 😄

pkeugine commented on src/main/kotlin/blackjack/model/CardDeck.kt:
```diff
@@ -0,0 +1,29 @@
+package blackjack.model
+
+class CardDeck(
```
**Q. Should we simulate multiple decks, like in a real casino? Right now, it's a single deck. Using more might better reflect how Blackjack is usually played.**

Since this decision is purely about the rule and flow about your blackjack implementation,
it's totally up to you to decide.

Personally, I think it'll be helpful for you to upgrade the game by using multiple decks,
since you'll be able to experience 'updates in project planning'. 👍

deniz-oezdemir replied:
I have decided on 4 sets of cards for the deck and changed the initialization logic in `CardDeck` accordingly. :-)

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -0,0 +1,98 @@
+import blackjack.view.OutputView.displayInitialCardsMessage
+
+class Controller {
+    private val deck = CardDeck()
+    private var _players: Players? = null
+    private val dealer = Dealer()
```
**Q. We keep some private variables in the Controller to track state. Is this an acceptable tradeoff, or should we use a more declarative approaches and pass variables between functions instead?**

Personally, I think it's fine in this level of project scale.

However, I want to know what that `tradeoff` is that you've thought of!
What are the pros and cons of this?

deniz-oezdemir replied:
The tradeoff for us in using private variables was that it's
- more readable, and
- less code for passing variables between functions
vs.
- more difficult to write tests
- more difficult implementing concurrency if needed

deniz-oezdemir replied:
Alright, then I'm leaving it as is for now. :-)

pkeugine replied:
Since this is a controller, I agree it might feel harder to write tests.

But if it were part of the domain layer, having private fields wouldn’t necessarily make testing harder than passing variables between functions. In fact, the approach of using fields and injecting them via constructors is called dependency injection. It makes testing much easier by letting you control the dependencies; definitely worth checking out when you have time! It’s actually simpler than it sounds. 🙂

Other than that, great tradeoff analysis, very helpful!

pkeugine commented on src/main/kotlin/blackjack/model/ResultEvaluator.kt:
```diff
@@ -0,0 +1,66 @@
+}
+
+data class PlayerResult(var win: Boolean = false, var draw: Boolean = false, val name: String = "") {
+    override fun toString(): String {
+        if (win) return "$name: win"
+        if (draw) return "$name: draw"
+        return "$name: loss"
```
**Q. We’ve overloaded toString() to make it easier to print results. Should we instead move formatting into a separate function inside OutputView?**

Yes, I strongly recommend moving this logic to `OutputView`, since it's tightly coupled to 'console based UI'. In different kind of view other than console, we might want to make our output different, hence putting each of those output logic in their own respectable classes.

As for `toString()`, it's mostly used for debugging or logging purposes since debugging and logging output don't change rather we use one kind of view or another.

deniz-oezdemir replied:
Moved all printing logic from `ResultEvaluator` into `OutputView`.
`toString()` now contains only neutral diagnostic info, while `OutputView` has `formatDealer` and `formatPlayer` for console output.

pkeugine commented on src/main/kotlin/blackjack/model/ResultEvaluator.kt:
```diff
@@ -0,0 +1,66 @@
+            else -> PlayerResult(draw = true, name = player.name)
+        }
+    }
+}
```
I noticed this red indicator at the end of some files.
It would be helpful to find out why having a new line at the end of the file is preferred.  
**Hint**: [it's related to missing empty lines](https://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-a-source-file) 😉  

In IntelliJ, there's an option to automatically add a newline at the end of a file. It’s worth checking that out.

deniz-oezdemir replied:
Thanks for the source! I'll add a newline at every file end from now on. :-)

pkeugine commented on src/main/kotlin/blackjack/model/Suit.kt:
```diff
@@ -0,0 +1,8 @@
+package blackjack.model
+
+enum class Suit(val suit: String) {
+    HEARTS("♥"),
+    DIAMONDS("♦"),
+    CLUBS("♣"),
+    SPADES("♠"),
```
I see you've added a visual effect for this game 👍 
Just like my previous comment, I personally think this visual effect also belongs to `OutputView`,
since there's a chance for us to use different way of showing the cards in other output strategy.

deniz-oezdemir replied:
Moved all card formatting logic from the model to the `OutputView` layer, so the model remains UI-agnostic.
Now `displayAllCardsMessage` and `displayFirstCardMessage` use helper methods in OutputView (`cardsToString` and `formatCard`) to build the card strings with suit symbols.

pkeugine commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,31 @@
+    fun promptForDraw(player: Player): Boolean {
+        println("\nWould ${player.name} like to draw another card? (y for yes, n for no)")
+        val input = readln().trim()
+        if (input.isEmpty()) {
+            throw IllegalArgumentException(ErrorMessage.EMPTY_INPUT.toString())
+        }
```
```suggestion
        require(input.isNotEmpty()) { ErrorMessage.EMPTY_INPUT.toString() }
```

We can utilize `require` for this logic!

deniz-oezdemir replied:
Thanks for the suggestion!

pkeugine commented on src/main/kotlin/blackjack/view/InputView.kt:
```diff
@@ -0,0 +1,31 @@
+        }
+    }
+
+    private const val YES = "y"
+    private const val NO = "n"
```
These constants made me think :  
* (This is my personal opinion) Do they really require names? I think it's common for anybody to think 'y' input is a yes, and 'n' input is a no. Naming them actually made it longer in code.

* The placement of these constants could be in the top of the object, since that is the most common convention. Or to make it simpler, since they are only used in one method, keep it inside that method.

deniz-oezdemir replied:
Thanks for the suggestion!
I moved the `YES` and `NO` constants to the top of the `InputView` objec.
While it’s true that "y" and "n" are intuitive, I find naming them improves code readability and reduces risk of typos.

pkeugine replied:
Nice, I respect your idea 💯 
Let's keep it that way :)

pkeugine commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    fun displayAllCardsMessage(participant: Participant) {
+        print(participant.name + "'s cards: ")
+        println(participant.cardsToString())
```
```suggestion
    println("${participant.name}'s cards: ${participant.cardsToString()}")

```

I think we can achieve the same output with less `println` method call by refactoring like above.
Also, if we need to print multiple lines, we can use multi-line string feature that Kotlin provides.

Why don't we use them to refactor our view related logic?

deniz-oezdemir replied:
Yes, I will try to reduce the calls to `println`.

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -0,0 +1,98 @@
+        repeat(MAX_TRIES){
+            try {
+                return InputView.promptForDraw(player)
+            } catch (e: IllegalArgumentException) {
+                println(e.message)
```
Although this is a simple logic, I think view related classes should take care of printing out errors as well. Give them the exception, messages, causes, or anything, and let them take care of printing what they want. How does this sound?

pkeugine replied:
If we extend this idea, repeating `MAX_TRIES` until we get the desired input can be done in view as well. Controller and domain related logics don't need to know how many tries were taken before we actually play blackjack. They just want to play the game with nice player names and their 'yes's and 'no's. :)

deniz-oezdemir replied:
I moved the input retry and error handling logic into the InputView methods.

Also, I moved all player name validation checks into a single `validatePlayerName` function, which is also used when reading user input.

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -0,0 +1,98 @@
+        printResults()
+    }
+
+    fun printResults() {
```
```suggestion
    private fun printResults() {
```

Let's keep methods as private if they aren't used by other classes.
Please check other methods across the whole project as well.

deniz-oezdemir replied:
Thanks for the suggestion! I made the `printResults()` method and a lot of other methods private. I also enabled visibility inspections in IntelliJ to help catch similar cases automatically and keep the code cleaner by restricting method visibility whenever possible.

pkeugine commented on src/main/kotlin/blackjack/controller/Controller.kt:
```diff
@@ -0,0 +1,98 @@
+        players.forEach { OutputView.displayParticipantStatus(it) }
+    }
+
+    fun Dealer.dealCards() {
+        while (this.shouldDraw()) {
+            OutputView.displayDealerDrawMessage()
+            this.drawCard(deck)
+        }
+    }
+
+    fun Players.dealCards() {
+        players.forEach { dealCards(it) }
+    }
```
Keeping these extension functions in between controller methods will make maintaining the project very difficult, since we're extending domain's logic in controller. Why don't we move these to somewhere else?

I do understand this happened because we want to print the status in these methods, but we'll have to take out `OutputView` logic from these domain logic to properly implement MVC pattern.

deniz-oezdemir replied:
You’re right, the extension functions were mixing domain behaviour with UI calls inside the controller. I removed the `Dealer.dealCards()` and `Players.dealCards()` extensions and added methods that replace those to the `Dealer` class.

pkeugine commented on src/test/kotlin/blackjack/DeckTest.kt:
```diff
@@ -0,0 +1,42 @@
+package blackjack
+
+import blackjack.model.Card
+import blackjack.model.CardDeck
+import blackjack.model.Rank
+import blackjack.model.Suit
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertDoesNotThrow
+import org.junit.jupiter.api.assertThrows
```
Let's not forget to check if there are unused imports after refactoring.

deniz-oezdemir replied:
Yes, I have set my IDE to automatic checking of imports. :-)

pkeugine commented on src/test/kotlin/blackjack/ResultsTest.kt:
```diff
@@ -0,0 +1,124 @@
+        val players = Players(listOf(player))
+        val evaluator = ResultEvaluator(players, dealer)
+        val results = evaluator.calculateResults()
+        OutputView.displayAllCardsMessage(dealer)
+        OutputView.displayAllCardsMessage(players.players[0])
```
Adding in `OutputView` in `ResultEvaluator` related  test seems unnecessary unless we use it as a debugging purpose. If it was used for debugging purpose, we can remove it after debugging, so that it doesn't get included in the commit.

If we wanted to make an integration test, there must be some kind of an assertion that checks if the `OutputView` behaved like we wanted it to.

Oh, and I think the name of this test class can be changed to `ResultEvaluatorTest`, since all the tests are about asserting it's logic.

deniz-oezdemir replied:
Yes, we will remove these method calls in the test cases.

Yes, we will rename the test class. :-)

pkeugine commented on src/test/kotlin/blackjack/TotalTest.kt:
```diff
@@ -0,0 +1,73 @@
+    fun `test if many aces are reduced`(){
+        val handCards = HandCards()
+        handCards.add(Card(Rank.ACE, Suit.DIAMONDS))
+        assertEquals(handCards.total, 11)
```
I want to know when you use `assertThat` and `assertEquals`. What are the differences between them? :)

deniz-oezdemir replied:
`assertThat(...).isEqual` is more readable and less prone to confusion of the `actual` and the `expected`. So generally we should probably use `assertThat` instead of `assertEquals`, right?

pkeugine replied:
Nice 👍 
I agree, and to top it off, `assertThat()` can chain bunch of neat methods, making it easier to assert things much conveniently.

deniz-oezdemir replied:
Thanks for the info. Then I will try to use mainly `assertThat()` in the future.

pkeugine commented on build.gradle.kts:
```diff
@@ -18,6 +18,7 @@ dependencies {
     testImplementation("org.junit.jupiter", "junit-jupiter", "5.10.2")
     testImplementation("org.assertj", "assertj-core", "3.25.3")
     testImplementation("io.kotest", "kotest-runner-junit5", "5.8.0")
+    testImplementation(kotlin("test"))
```
What are the reasons behind the inclusion of this dependency?
Also, I want you to tell me what this `kotlin("test")` does. 🙂

deniz-oezdemir replied:
It was included by mistake because we briefly included something like `kotlin.test.assertEquals` from that package but then removed it as JUnit supplies everything we need for testing. Forgot to remove the dependency. Will remove it. :-)

pkeugine reviewed (changes requested):
Hello Deniz,
I am doing well, and I hope you're good as well.
Nice work on blackjack mission! The code was very easy to read 👍 

I left some comments so it'll be nice for you to check them out before we move onto the next step.

If you have any questions or opinions, feel free to leave comments in this PR or hit me with a DM.

Happy hacking! 🎴 🎴 🎴

deniz-oezdemir committed:
Update src/main/kotlin/blackjack/view/InputView.kt

Co-authored-by: Eugine Park <pkeugine@gmail.com>

deniz-oezdemir committed:
Update src/main/kotlin/blackjack/view/OutputView.kt

Co-authored-by: Eugine Park <pkeugine@gmail.com>

deniz-oezdemir committed:
refactor(OutputView): minimize `println()` calls
- use kotlin string templates and string multiline features to make fewer `println()` calls

deniz-oezdemir committed:
fix(DSLStudy): delete incomplete study test file

deniz-oezdemir committed:
fix(gitignore): delete wrong include for study test file from gitignore

deniz-oezdemir committed:
refactor(Rules): extract blackjack rule constants

deniz-oezdemir committed:
refactor(CardDeck): change deck to consist of 4 set of cards
- refactor tests to account for the new size of the card deck

deniz-oezdemir committed:
refactor(OutputView, ResultEvaluator): move result formatting to OutputView
- Add formatDealer() and formatPlayer() helpers
- displayResults() uses new helpers and raw string block
- Simplify toString() in DealerResult and PlayerResult for logging only

deniz-oezdemir committed:
style(TotalTest): add newline at the end of the file

deniz-oezdemir committed:
refactor(OutputView): centralize card formatting in OutputView
- replace model calls with OutputView helpers cardsToString and formatCard
- keep suit symbols mapping inside OutputView for consistent UI rendering

deniz-oezdemir committed:
style(input): move YES/NO constants to top of InputView object

deniz-oezdemir committed:
refactor(InputVIew): move input retry and validation handling to InputView
- clean up controller by delegating input retry and error printing to InputView
- add validation for name constraints when taking user input

deniz-oezdemir committed:
fix(OutputView): show card ranks as symbols (K, Q, J, A) instead of numeric values again

deniz-oezdemir committed:
fix(Controller, HandCards, InputVIew, OutputView, ResultsEvaluator): set methods private if possible

deniz-oezdemir committed:
refactor(Controller, Dealer): replace extension functions with internal service methods
- Remove Dealer.dealCards() and Players.dealCards() extension functions
- Add serviceDealer(), servicePlayers(), and dealCards() to Dealer
- Refactor tests according to above changes

deniz-oezdemir committed:
refactor(ResultEvaluatorTest): remove OutputView calls and rename test class

deniz-oezdemir committed:
build: remove unused kotlin("test") dependency

deniz-oezdemir committed:
refactor(Players): remove unused property `size`
- Apply autoformatting with ktlintFormat

deniz-oezdemir commented:
Hi @pkeugine, thanks so much for your detailed review! 😄

I have addressed each comment and implemented changes accordingly.

There are still two unresolved change requests/suggestions where I would really appreciate your input before proceeding further. Your perspective would help clarify the best direction there.

I would really appreciate it if you could take another look whenever you have time. Let me know if there is anything else you’d like me to adjust. Thanks again, and I hope you have a great start to the new week! 🙏

deniz-oezdemir committed:
feature(DSL): Implement Kotlin DSL for Person creation with name, company, skills, and languages
- Add unit tests covering all DSL components and properties

deniz-oezdemir commented:
Hi @pkeugine,
I have just added a Kotlin DSL for constructing Person objects.
I have also added corresponding unit tests to verify each part of the DSL.
This was an additional task following one of the lectures and I would really appreciate if you could review it too.
Thanks, Deniz

pkeugine commented on src/test/kotlin/blackjack/DeckTest.kt:
```diff
@@ -0,0 +1,40 @@
+        assertEquals(deck1.size, deck2.size - 1)
+    }
+
+    @Test
+    fun `test draw more than deck`() {
+        val deck = CardDeck()
+        var card = Card(Rank.ACE, Suit.HEARTS)
+        repeat(208) {
+            card = deck.hit()
+        }
+        assertThrows<Exception> { card = deck.hit() }
+    }
```
I think we can simplify this test by :  
* remove `var card`. It doesn't look like it's necessary.
* rename it. For example : `throws when drawing more cards than deck contains` since it's asserting for an exception.
* specify the exception : For example : `assertThrows<NoSuchElementException>`. This will allow us to assert the logic with more precision.

Think about this in the next step :)

deniz-oezdemir replied:
You are right, I will refactor this test accordingly.

pkeugine commented on src/main/kotlin/study/DSL.kt:
```diff
@@ -0,0 +1,73 @@
+    private val softSkills = mutableListOf<String>()
+    private val hardSkills = mutableListOf<String>()
+
+    fun soft(value: String) {
+        softSkills += value
+    }
+
+    fun hard(value: String) {
+        hardSkills += value
+    }
```
```suggestion
    fun soft(vararg values: String) {
        softSkills += values
    }

    fun hard(vararg values: String) {
        hardSkills += values
    }
```

This kind of design would allow us to use it like :
```kotlin
soft("Teamwork", "Speech", "Leadership")
hard("Kotlin", "Gradle")
```

It's just a suggestion though. You did a nice job 👍

deniz-oezdemir replied:
Thanks for the suggestion! That’s a great idea that makes the DSL more flexible. I have updated the methods accordingly so multiple skills can now be added in a single call. Appreciate this feedback!

pkeugine commented on src/main/kotlin/study/DSL.kt:
```diff
@@ -0,0 +1,73 @@
+package study
```
It's not a must for now, but there's this thing called [`@DslMarker`](https://kotlinlang.org/api/core/kotlin-stdlib/kotlin/-dsl-marker/) that you can also use. It'll help you make a better DSL. Check it out when you have the time :)

deniz-oezdemir replied:
Thanks for the source, it seems like the `DslMarker` makes the DSL safer to use by restricting visibility in nested DSL scopes.

pkeugine commented on src/main/kotlin/blackjack/model/Participant.kt:
```diff
@@ -0,0 +1,23 @@
+
+    fun isBusted(): Boolean = (handCards.total > Rules.BLACKJACK_TARGET)
+
+    fun cardsToString(): String {
+        return handCards.cards.joinToString()
+    }
+
+    fun firstCardToString(): String {
+        return handCards.cards[0].toString()
+    }
```
Let's remove methods if they aren't used anywhere.

deniz-oezdemir replied:
Yes, I have realized this also while working on step2 and removed them. Good catch!

pkeugine commented on src/main/kotlin/blackjack/view/OutputView.kt:
```diff
@@ -0,0 +1,72 @@
+            else -> "${result.name}: loss"
+        }
+
+    // TODO: refactor Pair to value class?
+    fun displayResults(results: Pair<DealerResult, List<PlayerResult>>) {
```
It'll be nice if we could leave out `TODO`s from PR, and also keep them out of the main branch.
TODO's are mostly for local workspace, and should be implemented if we're trying to merge it to the main branch.

deniz-oezdemir replied:
Noted, TODOs are not to be pushed. :-)

pkeugine reviewed (approved):
Hello Deniz,
Nice work going through all the comments 👍 
You've refactored the code very cleanly, and I think we're ready to move on to the next step.

I've added some additional reviews, but they are minor.
Think about them while working on the next step, and we'll talk about it in the next PR if needed 👍 

If you have any questions or opinions, feel free to DM me.

Again, nice work,
and I'll see you soon!
Happy Hacking 🎴 🎴 🎴
